internal/fs/dir_handle.go

```diff
@@ -56,36 +55,63 @@ type dirHandle struct {
 	//
 	// GUARDED_BY(Mu)
 	entriesValid bool
+
+	//condition variable is for signalling whether a fresh set of entries has been fetched
+	cond *sync.Cond
+	//error during the fetching goroutine must be communicated using this to the main go routine
+	//serving kernel requests
+	err error
+
+	//using this as a identification flag to indicate all entries present in the directory
+	//have been fetched already so that the main goroutine does not wait indefinitely for more entries.
+	fetchOver bool
```

Comment:
in what scenario does main goroutine wait indefinitely?

Reply:
In the main goroutine, after the entries have been fetched completely, the other goroutine will terminate.But since main go routine does not have knowledge of it. it will ask for the next entries Eg. 5k objects have been fetched and no more left to fetch.
Since the offset 5001 > len of entries, the main go routine will see it as a fetch is in process and will wait for a signal but since there is no broadcast from the other goroutine which has already exited, it will continue to be blocked.